### PR TITLE
Redesign and improve reporter error display

### LINKED ISSUE #6722: Assertion diff is not shown when an assertion fails and is retried
### Current behavior:

When an assertion is retried (using `cy.should()`) and fails, it does not show a diff in the stdout. If the same assertion is not retried, it will show the diff.

### Desired behavior:

Failing assertions should show diffs in the stdout regardless of whether they are retried or not.

### Test code to reproduce

This works properly and shows what the diff should look like.

```javascript
it('non-dom - fails without retry', () => {
  expect([]).to.eql([1, 2, 3])
})
```

Screen Shot 2020-03-13 at 3 07 18 PM

This demonstrates the issue.

```javascript
it('non-dom - fails with retry', () => {
  cy.wrap([]).should('eql', [1, 2, 3])
})
```

Screen Shot 2020-03-13 at 3 17 17 PM

### Versions

Cypress 4.1.0

### LINKED ISSUE #6723: Diff should not be shown when expected or actual is a dom element
### Current behavior:

When an assertion involving a dom element retried fails, it shows a diff that includes an ugly dom object.

Screen Shot 2020-03-13 at 3 21 42 PM
Screen Shot 2020-03-13 at 3 21 54 PM

Note that this doesn't happen if the assertion is retried (with `cy.should()`).

### Desired behavior:

Failing dom assertions should not show the diff, and not show the ugly dom object.

### Test code to reproduce

```javascript
it('dom - fails without retry', () => {
  cy.get('body').then(($body) => {
    expect($body).to.have.class('foo')
  })
})
```

### Versions

Cypress 4.1.0

### PR DESCRIPTION
**Pull request by chrisbreiding**
- Closes #6722
- Closes #6723

Partial work originally pulled from #3930 

- Add markdown support
- Collapse stacktrace
- Separate docs url and add link in reporter
- Refactor error utils

### PR Tasks

- [x] Have tests been added/updated?
- [x] Has the original issue been tagged with a release in ZenHub? 
- [x] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- N/A Have API changes been updated in the [`type definitions`](cli/types/index.d.ts)?
- N/A Have new configuration options been added to the [`cypress.schema.json`](cli/schema/cypress.schema.json)?

### CODE REVIEW

File: packages/server/__snapshots__/1_caught_uncaught_hook_errors_spec.coffee.js [hunk 1]
```diff
@@ -40,21 +40,25 @@ exports['e2e caught and uncaught hooks errors failing1 1'] = `
...
   1) s1a "before each" hook for "t2a":
-     CypressError: Timed out retrying: Expected to find element: '.does-not-exist', but never found it.
 
-Because this error occurred during a 'before each' hook we are skipping the remaining tests in the current suite: 's1a'
+      AssertionError: Timed out retrying: Expected to find element: \`.does-not-exist\`, but never found it.
+
+Because this error occurred during a \`before each\` hook we are skipping the remaining tests in the current suite: 's1a'
+      + expected - actual
+
+
```
Comment (brian-mann) [inline packages/server/__snapshots__/1_caught_uncaught_hook_errors_spec.coffee.js]: investigate whether this change is intentional

File: packages/driver/src/cy/retries.coffee [hunk 2]
```diff
@@ -58,15 +59,22 @@ create = (Cypress, state, timeout, clearTimeout, whenStable, finishAssertions) -
...
+        prependMsg = $errUtils.errMsgByPath("miscellaneous.retry_timed_out")
+
+        retryErrProps = $errUtils.modifyErrMsg(error, prependMsg, (msg1, msg2) ->
+          return "#{msg2}#{msg1}"
+        )
+
+        retryErr = $errUtils.mergeErrProps(error, retryErrProps)
+
+        if retryErr.type is "existence"
+          retryErr.showDiff = false
```
Comment (brian-mann) [inline packages/driver/src/cy/retries.coffee]: TODO: brian - need to find a better place for this - likely in `chai.js`

File: packages/server/__snapshots__/6_uncaught_spec_errors_spec.coffee.js [hunk 3]
```diff
@@ -371,7 +371,8 @@ exports['e2e uncaught errors / failing5'] = `
       [stack trace lines]
 
   4) foo quux2 fails:
-     ReferenceError: foo is not defined
+     ReferenceError: foo is not defined (and Mocha's done() called multiple times)
+  ReferenceError: foo is not defined
```
Comment (brian-mann) [inline packages/server/__snapshots__/6_uncaught_spec_errors_spec.coffee.js]: this does not look right

File: packages/server/__snapshots__/3_issue_674_spec.coffee.js [hunk 4]
```diff
@@ -26,15 +26,13 @@ exports['e2e issue 674 / fails'] = `
...
+     Because this error occurred during a \`before each\` hook we are skipping the remaining tests in the current suite: 'issue 674'
+  rror
       [stack trace lines]
 
   2) issue 674 "after each" hook for "doesn't hang when both beforeEach and afterEach fail":
-     Error: 
-
-Because this error occurred during a 'after each' hook we are skipping the remaining tests in the current suite: 'issue 674'
+     Because this error occurred during a \`after each\` hook we are skipping the remaining tests in the current suite: 'issue 674'
+  rror
```
Comment (chrisbreiding) [inline packages/server/__snapshots__/3_issue_674_spec.coffee.js]: This doesn't look right

File: packages/driver/src/cy/commands/connectors.coffee [hunk 5]
```diff
@@ -359,6 +358,9 @@ module.exports = (Commands, Cypress, cy, state, config) ->
               Yielded: getFormattedElement(value)
             })
 
+            if traversalErr
+              obj.Error = "#{traversalErr.name}: #{traversalErr.message}"
```
Comment (chrisbreiding) [inline packages/driver/src/cy/commands/connectors.coffee]: Why is this necessary?

**Pull request merged**